### PR TITLE
fix create view on exist table throw view exist exception

### PR DESCRIPTION
**Pull request by kyo-tom**
fix hive connector create view on exist table throw view exist exception. #10037 
before:
```text
trino:default> CREATE VIEW one AS SELECT * FROM one;
Query 20211206_045554_00007_33tp9 failed: View already exists: 'default.one'
```
after:
```text
trino:default> CREATE VIEW one AS SELECT * FROM one;
Query 20211206_045554_00007_33tp9 failed: Table already exists: 'default.one'
```

Fixes https://github.com/trinodb/trino/pull/10186

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 1]
```diff
@@ -2209,7 +2209,12 @@ public void createView(ConnectorSession session, SchemaTableName viewName, Conne
         Optional<Table> existing = metastore.getTable(identity, viewName.getSchemaName(), viewName.getTableName());
         if (existing.isPresent()) {
             if (!replace || !isPrestoView(existing.get())) {
-                throw new ViewAlreadyExistsException(viewName);
+                if (TableType.VIRTUAL_VIEW.name().equals(existing.get().getTableType())) {
```
Comment (lukasz-stec) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2212]: IMO we should move this logic to the `CreateViewTask` and handle it in a similar way e.g. `RenameViewTask` handles it (including a check for materialized view). It would work then for all connectors (looks like iceberg connector may have the same issue) + would support MVs.

File: core/trino-main/src/test/java/io/trino/execution/TestCreateViewTask.java [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        public Optional<MaterializedViewDefinition> getMaterializedView(Session session, QualifiedObjectName viewName)
+        {
+            return Optional.empty();
+        }
+
+        @Override
+        public void createView(Session session, QualifiedObjectName viewName, ViewDefinition definition, boolean replace)
+        {
+            SchemaTableName viewSchemaTableName = viewName.asSchemaTableName();
+            if (tables.containsKey(viewSchemaTableName)) {
```
Comment (lukasz-stec) [inline core/trino-main/src/test/java/io/trino/execution/TestCreateViewTask.java:309]: the tests above are mostly checking this (mock) logic. If the checks are moved to CreateTableTask (see comment below) this will be unnecessary and you can extend the test class from BaseDataDefinitionTaskTest.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 1 above)
Comment (kyo-tom) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2212]: Yes, I also think it should work for all connectors, so the one that really needs to be enhanced is CreateViewTask. So I will resubmit a pr to make an enhancement to CreateViewTask, can I close this pr now?

Comment (lukasz-stec) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2212]: I think it's ok to make changes in this PR (you can squash commits if you need to)

Comment (kyo-tom) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2212]: cc @lukasz-stec , I am very confused about the function of ConnectorMetadata#getTableHandle in hive.it should return null when the relation is views (just like the comment says in this function). But if I create a view, it will not return null in hive because of tableActions will add a table when view create

Comment (lukasz-stec) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2212]: yeah, hive connector implementation of `ConnectorMetadata#getTableHandle` breaks the contract. I don't think it's that easy to fix though (some features may depend on this behavior)

Comment (kyo-tom) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2212]: IMO we can first judge whether the view exists according to Metadata#isView, if the view does not exist, then judge whether the table exists, we also need to verify whether this is feasible on other connectors. Do you think we need open other issus about hive connector implementation of ConnectorMetadata#getTableHandle breaks the contract.

Comment (lukasz-stec) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:2212]: > IMO we can first judge whether the view exists according to Metadata#isView

yes + `metadata.isMaterializedView` to check if its MV
> Do you think we need open other issus about hive connector implementation of ConnectorMetadata#getTableHandle breaks the contract.

Good question. There is this ticket https://github.com/trinodb/trino/issues/8349 which is a kind of prerequisite for fixing the hive connector but maybe we need another one.

Comment (kyo-tom) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java]: Hello, @lukasz-stec .Could you review my code again？ Thank you！

File: core/trino-main/src/test/java/io/trino/execution/TestCreateViewTask.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        QualifiedObjectName tableName = qualifiedObjectName("mock_table");
+        metadata.createTable(testSession, CATALOG_NAME, someTable(tableName), false);
+    }
+
+    @Test
+    public void testCreateViewNotExistsTrue()
+    {
+        QualifiedObjectName viewName = qualifiedObjectName("new_view");
+        getFutureValue(executeCreateView(asQualifiedName(viewName), false));
+        assertEquals(metadata.getCreateViewCallCount(), 1);
```
[the remark this answers is not in the document]
Comment (kyo-tom) [reply core/trino-main/src/test/java/io/trino/execution/TestCreateViewTask.java:86]: good idea

File: core/trino-main/src/test/java/io/trino/execution/BaseDataDefinitionTaskTest.java [hunk 4]
```diff
@@ -326,5 +326,20 @@ public void renameMaterializedView(Session session, QualifiedObjectName source,
...
+
+        public int getCreateViewCallCount()
+        {
+            return views.size();
+        }
+
+        public int getCreateMaterializedViewCallCount()
+        {
+            return materializedViews.size();
+        }
```
[the remark this answers is not in the document]
Comment (kyo-tom) [reply core/trino-main/src/test/java/io/trino/execution/BaseDataDefinitionTaskTest.java:343]: Yes,it can be revert if we use `metadata.isView` to check create view

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 5]
```diff
@@ -76,6 +80,17 @@ public String getName()
...
         accessControl.checkCanCreateView(session.toSecurityContext(), name);
 
+        if (metadata.isView(session, name) && !statement.isReplace()) {
+            throw semanticException(ALREADY_EXISTS, statement, "View '%s' already exists", name);
+        }
+        if (metadata.isMaterializedView(session, name)) {
+            throw semanticException(ALREADY_EXISTS, statement, "Materialized view '%s' already exists", name);
+        }
+        Optional<TableHandle> tableHandle = metadata.getTableHandle(session, name);
+        if (tableHandle.isPresent()) {
```
[the remark this answers is not in the document]
Comment (kyo-tom) [reply core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:90]: sorry, @lukasz-stec . I don't understand what you mean, , can you elaborate more？

[the remark this answers is not in the document]
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:90]: Sorry, I should say *may* break...

So because `getTableHandle` will return not empty `TableHandle` for a view in hive connector, when `statement.isReplace() == true`  (i.e. `CREATE OR REPLACE VIEW` was used) the method will trow
`throw semanticException(TABLE_ALREADY_EXISTS, statement, "Table '%s' already exists", name);` and it should replace the view definition successfuly.

[the remark this answers is not in the document]
Comment (kyo-tom) [reply core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java]: Thank you for your reminder, I still need to check if it is view when `statement.isReplace() == true`.Sorry ,can @lukasz-stec  you review again.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 6]
```diff
@@ -413,7 +413,7 @@ public void testReferencedTablesWithMaterializedViews()
     public void testReferencedTablesInCreateView()
             throws Exception
     {
-        runQueryAndWaitForEvents("CREATE VIEW mock.default.test_view AS SELECT * FROM nation", 2);
+        runQueryAndWaitForEvents("CREATE OR REPLACE VIEW mock.default.test_view AS SELECT * FROM nation", 2);
```
Comment (lukasz-stec) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:416]: is this change needed?

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 7]
```diff
@@ -76,6 +80,19 @@ public String getName()
...
+        if (metadata.isMaterializedView(session, name)) {
+            throw semanticException(ALREADY_EXISTS, statement, "Materialized view already exists: '%s'", name);
+        }
+        Optional<TableHandle> tableHandle = metadata.getTableHandle(session, name);
+        // Normally, when a table with the same name does not exist but a view with the same name exists
+        // tableHandle should return empty. But some connectors are exceptions, eg. hive connector...
+        if (tableHandle.isPresent() && !metadata.isView(session, name)) {
+            throw semanticException(TABLE_ALREADY_EXISTS, statement, "Table already exists: '%s'", name);
+        }
+
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:95]: this generally looks good to me.
One small comment is that we could cache
`metadata.isView(session, name)` result in the local variable and then skip the `isMaterializedView` and `getTableHandle` calls if it's true (i.e. if it's a view it is definitely not a materialized view nor a table). This would both avoid unnecessary overhead and allow you to drop the view check in the `tableHandle.isPresent()` if statement.

**Review by lukasz-stec — APPROVED** [review]
lgtm % comments

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 6 above)
Comment (kyo-tom) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:416]: Because the view named test_view  is add to ``MockConnectorMetadata`` by withGetViews in ``TestEventListenerBasic``, so ``metadata.isViews`` will always return true and throw a exception

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java (hunk 7 above)
Comment (kyo-tom) [reply core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:95]: I will do it

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 8]
```diff
@@ -76,6 +80,22 @@ public String getName()
 
         accessControl.checkCanCreateView(session.toSecurityContext(), name);
 
+        boolean isView = metadata.isView(session, name);
+        if (isView && !statement.isReplace()) {
+            throw semanticException(ALREADY_EXISTS, statement, "View already exists: '%s'", name);
+        }
+        else if (metadata.isMaterializedView(session, name)) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:87]: Change the order:
* MV
* view
* table
This is the metadata check order that we use in other places. Let's be consistent here

Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:87]: `else` is not needed

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 9]
```diff
@@ -76,6 +80,22 @@ public String getName()
 
         accessControl.checkCanCreateView(session.toSecurityContext(), name);
 
+        boolean isView = metadata.isView(session, name);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:83]: Check `isView` only after `isMaterializedView` (see other comment below)

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 10]
```diff
@@ -76,6 +80,22 @@ public String getName()
...
+        if (isView && !statement.isReplace()) {
+            throw semanticException(ALREADY_EXISTS, statement, "View already exists: '%s'", name);
+        }
+        else if (metadata.isMaterializedView(session, name)) {
+            throw semanticException(ALREADY_EXISTS, statement, "Materialized view already exists: '%s'", name);
+        }
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:89]: I think the check should be:
```
if (metadata.isMeterializedView(...)) {
  throw
}
if (metadata.isView()) {
  if (!statement.isReplace()) {
    throw 
  }
// do not pull table handle if we know object is a view
} else if (metadata.getTableHandle(session, name).isPresent()) {
  throw
}
```

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java (hunk 8 above)
Comment (kyo-tom) [reply core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java]: @sopel39 I had changed the order,can you reviewer again

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 11]
```diff
@@ -76,6 +79,18 @@ public String getName()
 
         accessControl.checkCanCreateView(session.toSecurityContext(), name);
 
+        if (metadata.isMaterializedView(session, name)) {
+            throw semanticException(ALREADY_EXISTS, statement, "Materialized view already exists: '%s'", name);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:83]: use `TABLE_ALREADY_EXISTS`

File: core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java [hunk 12]
```diff
@@ -76,6 +79,18 @@ public String getName()
...
+        }
+        if (metadata.isView(session, name)) {
+            if (!statement.isReplace()) {
+                throw semanticException(ALREADY_EXISTS, statement, "View already exists: '%s'", name);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/CreateViewTask.java:87]: use `TABLE_ALREADY_EXISTS`

File: plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java [hunk 13]
```diff
@@ -556,7 +556,7 @@ public void testViews()
         assertUpdate("CREATE OR REPLACE VIEW test_view AS " + query);
 
         assertQueryFails("CREATE TABLE test_view (x date)", "View \\[default.test_view] already exists");
-        assertQueryFails("CREATE VIEW test_view AS SELECT 123 x", "View already exists: default.test_view");
+        assertQueryFails("CREATE VIEW test_view AS SELECT 123 x", ".*View already exists: 'memory.default.test_view'");
```
Comment (sopel39) [inline plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java:559]: Do we need `.*` as a prefix?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 6 above)
Comment (sopel39) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:416]: > Because the view named test_view is add to MockConnectorMetadata by withGetViews in TestEventListenerBasic, so metadata.isViews will always return true and throw a exception

Should use use different view name here then?

File: plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java (hunk 13 above)
Comment (kyo-tom) [reply plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java:559]: Yes, SemanticException will extractLocation like ``line 1:1: View already exists: 'memory.default.test_view'``

File: core/trino-main/src/test/java/io/trino/execution/TestCreateViewTask.java [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @Test
+    public void testCreateViewOnViewIfExists()
+    {
+        QualifiedObjectName viewName = qualifiedObjectName("existing_view");
+        metadata.createView(testSession, viewName, someView(), false);
+
+        assertTrinoExceptionThrownBy(() -> getFutureValue(executeCreateView(asQualifiedName(viewName), false)))
+                .hasErrorCode(ALREADY_EXISTS)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/execution/TestCreateViewTask.java:86]: `ALREADY_EXISTS` -> `TABLE_ALREADY_EXISTS`

File: core/trino-main/src/test/java/io/trino/execution/TestCreateViewTask.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @Test
+    public void testCreateViewOnMaterializedView()
+    {
+        QualifiedObjectName viewName = qualifiedObjectName("existing_materialized_view");
+        metadata.createMaterializedView(testSession, viewName, someMaterializedView(), false, false);
+
+        assertTrinoExceptionThrownBy(() -> getFutureValue(executeCreateView(asQualifiedName(viewName), false)))
+                .hasErrorCode(ALREADY_EXISTS)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/execution/TestCreateViewTask.java:129]: `TABLE_ALREADY_EXISTS`

File: plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java [hunk 16]
```diff
@@ -556,7 +556,7 @@ public void testViews()
         assertUpdate("CREATE OR REPLACE VIEW test_view AS " + query);
 
         assertQueryFails("CREATE TABLE test_view (x date)", "View \\[default.test_view] already exists");
-        assertQueryFails("CREATE VIEW test_view AS SELECT 123 x", "View already exists: default.test_view");
+        assertQueryFails("CREATE VIEW test_view AS SELECT 123 x", "line 1:1: View already exists: 'memory.default.test_view'");
```
Comment (sopel39) [inline plugin/trino-memory/src/test/java/io/trino/plugin/memory/TestMemoryConnectorTest.java:559]: `line 1:1:` -> this might be flaky. `.*` was better in this case

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 17]
```diff
@@ -413,13 +413,13 @@ public void testReferencedTablesWithMaterializedViews()
     public void testReferencedTablesInCreateView()
             throws Exception
     {
-        runQueryAndWaitForEvents("CREATE VIEW mock.default.test_view AS SELECT * FROM nation", 2);
+        runQueryAndWaitForEvents("CREATE VIEW mock.default.create_view_not_exist AS SELECT * FROM nation", 2);
```
Comment (sopel39) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:416]: `create_view_not_exist` -> `another_test_view` or something similar

**Review by sopel39 — APPROVED** [review]

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 17 above)
Comment (kyo-tom) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:416]: the view name need start with ``create`` because of withGetTableHandle in mock,so can we use ``create_another_test_view``?

**Comment by sopel39** [comment]
Fixes https://github.com/trinodb/trino/pull/10186

**Comment by sopel39** [comment]
merged thanks!
